### PR TITLE
Improve support for suffixed packages

### PR DESCRIPTION
**Pull request by abn**
* [x] I have added an entry to `docs/changelog.md`

## Summary of changes
This change allows for safe listing and upgrade of packages installed with a suffix. As part of this change, a new metadata property "suffix" has been introduced.

## Test plan

Tested by running
```sh
pipx install --suffix=_test pycowsay
pipx upgrade pycowsay_test
pipx inject pycowsay_test pip
pipx list
pipx uninstall pycowsay_test
```

### CODE REVIEW

File: src/pipx/pipx_metadata_file.py
Comment (itsayellow) [inline src/pipx/pipx_metadata_file.py:41]: For any change to the pipx metadata format, line 66 `self._pipx_metadata_version: str = "0.1"` needs to be updated with an incremented version string.

I think a good approach would be a new module global `PIPX_METADATA_VERSION = "0.2"` at the top, and change line 66 to `self._pipx_metadata_version: str = PIPX_METADATA_VERSION`

Comment (abn) [reply src/pipx/pipx_metadata_file.py:41]: @itsayellow I have made the changes but using a slightly different approach. I reckon since this version is tied to the class, I have made it a class variable. Should ahve the same desired effect. Happy to change it to your suggestion if you think that is better.

Comment (itsayellow) [reply src/pipx/pipx_metadata_file.py:41]: Can you use something more like `__METADATA_VERSION__` instead of `__VERSION__`?  I want to make sure nobody confuses this with the version of pipx, it is specifically for the format of the metadata we write to file.

Comment (abn) [reply src/pipx/pipx_metadata_file.py:41]: @itsayellow consider it done! :)

**Comment by abn** [comment]
@itsayellow is there any further changes required for this one?

**Comment by itsayellow** [comment]
> @itsayellow are there any further changes required for this one?

Still working on it.  With unpaid volunteer labor you need to be patient. 🙂

**Comment by abn** [comment]
> Still working on it. With unpaid volunteer labor you need to be patient.

Well aware of those pains, was not intending to push for resolution. Just making sure to dot the i's and cross the t's. :)

**Comment by itsayellow** [comment]
Oh my gosh, so currently it looks like parts of pipx are actually broken when using packages with a suffix?  I had no idea.  This is less of an "improvement" and more of a necessary bugfix!

**Comment by cs01** [comment]
@abn Thank you for the PR. I looked it over and it looks good to me -- I have no comments. It is rare to see no lint errors, unit tests passing AND new unit tests all in a single commit. Impressive! 

I'll wait for @itsayellow's review before merging since he wrote the metadata code and knows it better than me.

**Comment by cs01** [comment]
> Oh my gosh, so currently it looks like parts of pipx are actually broken when using packages with a suffix? I had no idea. This is less of an "improvement" and more of a necessary bugfix!

Yeah I was kind of surprised that these changes were needed, though I'm glad they were; they seem like basic features of the suffix functionality.

**Comment by abn** [comment]
@itsayellow ah yes, it is broken. I figured you were aware of this ... limitation of the current implementation. I recently started using pipx quite regularly and is finding the suffix feature rather useful for triage/testing the cli tool I help out with. That is what initiated the fix. :)

@cs01 I think I had a lint failire on my inital push ... so not all clean :wink:

File: src/pipx/commands/common.py
Comment (itsayellow) [inline src/pipx/commands/common.py:152]: I think instead of these two line changes, we need to go a little farther back and change starting line 141:
```
if package is None:
    package = path.name
```
to instead be
```
if package is None:
  package = venv.package_metadata.main_package["package"]
```
Since we have now broken old pipx's one-to-one relationship between path <-> main_package name.

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: And then I think you don't need `venv.get_package_metadata()` and you don't need to search for the package name with its for-loop.

File: src/pipx/commands/common.py
Comment (itsayellow) [inline src/pipx/commands/common.py:179]: I think this should not be necessary if you get `package` correct in the first place (see above comment.)

File: src/pipx/commands/common.py
Comment (itsayellow) [inline src/pipx/commands/common.py:149]: In the function definition for get_package_summary you no longer have any use for argument `suffix` and it should be removed.

**Review by itsayellow — CHANGES_REQUESTED** [review]
In general I'd like to look over this again after the following change that I've alluded to in the comments:

I think you can remove the uses of `package_metadata = venv.get_package_metadata()` and revert back to `package_metadata = venv.package_metadata[package]` if you first make sure you get the correct `package`, which is now not necessarily the name of the directory.  The main_package name is already in the metadata and you should use that (as I've marked up in the source) instead of starting out assuming that the directory path is the package name.

Let me know if there's something I'm missing and this isn't possible in all cases, but from my read I think what I said can be done.

File: src/pipx/commands/common.py
Comment (abn) [reply src/pipx/commands/common.py:152]: I am not sure if this will cover all the cases unfortunately. I had something similar in my first attempt. The issue was upgrade of injected packages that failed in this scenario. Additioanlly, the `get_package_metadata` is also useful if the suffix for installed app packages are enabled int he future. While it is probably not the most efficient method, it does the trick reliably. I am happy to make the suggested changes, and handle the key error explicitly in the upgrade code. Unless there is a cleaner way of doing this that I am missing.

**Comment by abn** [comment]
> Let me know if there's something I'm missing and this isn't possible in all cases, but from my read I think what I said can be done.
@itsayellow  see my response for why I ended up doing that. TLDR; upgrade of injected packages was not working as expected.

File: src/pipx/commands/common.py
Comment (itsayellow) [reply src/pipx/commands/common.py:152]: Can you explain what's going wrong with upgrade?  I wasn't aware we could upgrade injected packages.

Also, could you explain what you mean by "useful if the suffix for installed app packages are enabled int he future"?

In general I'd like to fix the implicit assumption in the code that pathname==main_package, because ever since we have suffixes that is incorrect.

Comment (abn) [reply src/pipx/commands/common.py:152]: I am purely going by the possible code-paths here. It might be that I am misunderstanding what `pipx` wants to support. In theory any package in a venv ca be updated based on this signature.

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L15-L23

Replacing the following with `package_metadata = venv.pipx_metadata.main_package` seems incorrect.

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L42

And following that there is also this call, which I expect is intended to expose apps (my understanding of the pipx app concept might be incomplete) from a pipx environment. 

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L70-L72

Since each of the packages now have a `suffix` property, in theory each exposed app can use a suffix. So this implies that if required, one can expose each app with its own suffix or an inherited suffix. This might not be what the project wants, and I can appreciate that.

I feel that in order to break that assumption safely old installations in the wild will need to be updated I feel. Otherwise a `pipx` upgrade can break existing environments. Again, a fair few assumptions here.

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: Currently we don't support upgrading anything but the main package (see the line `# TODO 20191026: upgrade injected packages also (Issue #79)`)

> Replacing the following with `package_metadata = venv.pipx_metadata.main_package` seems incorrect.

That would be incorrect.  What I was proposing was to obtain metadata on the main package was:
```
package = venv.pipx_metadata.main_package
package_metadata = venv.package_metadata[package]
```

> Since each of the packages now have a suffix property, in theory each exposed app can use a suffix. So this implies that if required, one can expose each app with its own suffix or an inherited suffix. This might not be what the project wants, and I can appreciate that.
>
> I feel that in order to break that assumption safely old installations in the wild will need to be updated I feel. Otherwise a `pipx` upgrade can break existing environments. Again, a fair few assumptions here.

I think right now we don't want to support any suffixes besides the main package, (@cs01 can correct me).  Doing otherwise sounds too complicated and confusing.  Currently the only pipx command with a `--suffix` option is `install`, so only the main package can currently have a suffix, and we don't have to worry about injected packages or dependencies in the wild having a suffix.

Definitely you are correct, `upgrade`'s use of `expose_apps()` needs a `suffix` argument, but I would argue that we currently only suffix the apps from the main package (as is the current behavior.)

In general, so far in pipx, injected packages are definitely second-class citizens.  We are not implementing everything for them that we do for the main package, like suffixes on their apps.

Grafting the `--suffix` option onto pipx clearly had more side-effects than originally envisaged.  I'm just trying to figure out the least-bad way to leave the code after still fixing these bugs.  Viewing all the code paths at this point in time as possible may not make sense because we probably need some refactoring and rethinking to completely integrate suffixes.

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: Did you actually see a failure with `upgrade` and `package = venv.pipx_metadata.main_package`, or are you basing this based just on the code?

Comment (abn) [reply src/pipx/commands/common.py:152]: > That would be incorrect. What I was proposing was to obtain metadata on the main package was:

That was a typo on my side. My point still stands that this overrides a mandatory argument in the `upgrade` function. Even when a package is provided, we reset this to the `main_package`. If this is indeed the intended behaviour we should probably drop the `package` parameter and simply rely on the metadata of the `Venv`.

> I would argue that we currently only suffix the apps from the main package (as is the current behavior.)

I am happy to do that, but I just find that a step backwards, especially since there already is a TODO to support this.

> Viewing all the code paths at this point in time as possible may not make sense because we probably need some refactoring and rethinking to completely integrate suffixes.

I agree. However, this is not necessarily handling all code paths, but rather just the ones that seemed obvious due to the interface. And as I said, happy to reduce the scope of the PR.

Comment (abn) [reply src/pipx/commands/common.py:152]: > package = venv.pipx_metadata.main_package
> package_metadata = venv.package_metadata[package]

Isin't `venv.pipx_metadata.main_package` a `PackageInfo` instance?

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: > > package = venv.pipx_metadata.main_package
> > package_metadata = venv.package_metadata[package]
> 
> Isin't `venv.pipx_metadata.main_package` a `PackageInfo` instance?

Ah yes you are right.  Sorry I was just typing from my mind.

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: As for `upgrade()`, right now the `venv_dir` and and `package` arguments I believe are more of an implementation detail that are supposed to track each other--`venv_dir` is just the full path and `package` is just the name of the package.

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: Another possibility that could be considered is that 
`@property def package_metadata(self)` in class `Venv` returns a dict with package + package_suffix as the keys to the dict instead of just package.  Do you have any idea if that would help things?  Or would it mess up something else that depends on unsuffixed package names?

Comment (abn) [reply src/pipx/commands/common.py:152]: I recall I had tried to use that as a fix initially; but there was something that felt harder than it should be and existing tests were failing. I can try that again see how far I can get with it. This simply avoids the loop in the cases where the main package is suffixed and is the one for which the metadata is being retrived. Although, overall I think it is probably not ideal as most lookups are package names, not name + suffix.

Additionally, while trying to implement your suggestions, I realiased that the `run_post_install_actions` will also have similar issues. SInce this is also called after inject with a non-main package. So it should handle both cases to support today's use cases. 

I could do something like this in places where a package is passed in. But then again this logic might be better placed in a method under `Venv`.

```py
    suffixed_package_name = f"{venv.pipx_metadata.main_package.package}{venv.pipx_metadata.main_package.suffix}"
    if package == suffixed_package_name:
        package = venv.pipx_metadata.main_package.package or package

    if package not in venv.package_metadata:
        raise PipxError(f"Package {red(bold(package))} not found in {venv.root}.")
```

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: What if you modified `def package_metadata(self)` to only apply the suffix for the main_package?  The main package is already applied in a separate if-clause anyway.

Is there a way this breaks?

Comment (itsayellow) [reply src/pipx/commands/common.py:152]: I was playing around with the code a bit, and following seems to work at least with `pipx list`:
```
    @property
    def package_metadata(self) -> Dict[str, PackageInfo]:
        return_dict = self.pipx_metadata.injected_packages.copy()
        if self.pipx_metadata.main_package.package is not None:
            return_dict[
                self.pipx_metadata.main_package.package + self.pipx_metadata.main_package.suffix
            ] = self.pipx_metadata.main_package
        return return_dict
```
I used most of your `pipx_metadata_file.py` code.

I don't think we should try to apply the suffix to everything, just the main package and its apps.

File: src/pipx/pipx_metadata_file.py
Comment (itsayellow) [inline src/pipx/pipx_metadata_file.py:38]: I think `suffix` should be a member of `PipxMetadata`.  I don't think we want more than one suffix for a given venv.

**Comment by itsayellow** [comment]
@cs01, there's some policy questions here that I'm curious for your thoughts about.
My thoughts:
* Only one suffix per venv
* Only the main package and its apps get the suffix (not injected packages + their apps)
* Should `--include-deps` apps also get the suffix?

File: src/pipx/commands/common.py
Comment (abn) [reply src/pipx/commands/common.py:152]: I need to double check how existing installs would be affected. Will do that when I get some time to look at this again. 

I suppose the policy question, even if not permanent, should be addressed before reducing the change scope.

**Comment by abn** [comment]
Assuming that the included apps are exposed to path, I'd suggest we should allow suffix configuration there too. 

I do currently have a use case where a certain pip version is a dependency that gets injected and would like to expose multiple versions from different environments. But this suffix might differ from the main package suffix. I appreciate this is not in scope for the PR, but wanted to mention this for the policy decision.

**Comment by itsayellow** [comment]
> I do currently have a use case where a certain pip version is a dependency that gets injected and would like to expose multiple versions from different environments. But this suffix might differ from the main package suffix. I appreciate this is not in scope for the PR, but wanted to mention this for the policy decision.

In this case, why would the pip suffix need to differ from the main package suffix?

**Comment by itsayellow** [comment]
My rationale for a simple policy regarding suffixes *right now*, is that it's already rippling through the code, and absorbing a simple pervasive change now and incrementally improving/expanding it later will be easier (and hopefully more successful) for implementation.  Also, a more complicated system of suffixes would be hard to explain/document from a user perspective.

I am sensitive to how useful a more powerful/expansive approach of suffixes could be, but there is a tradeoff between power and complexity.

**Comment by cs01** [comment]
> @cs01, there's some policy questions here that I'm curious for your thoughts about.
> My thoughts:
> 
> * Only one suffix per venv
> * Only the main package and its apps get the suffix (not injected packages + their apps)
> * Should `--include-deps` apps also get the suffix?

> Also, a more complicated system of suffixes would be hard to explain/document from a user perspective.

We need to consider first whether the feature can be understood and supported by part time volunteers, and second if a significant portions of users would use the feature.

pipx has been slowly accreting new features, and each new feature adds significant complexity. Between inject, `--include-deps`, and metadata tracking, the codebase has become much more complicated. pipx is, IMO, very useful as it is, and this suffix feature, although pretty cool, is a bit of a corner case.

We should not sacrifice maintainability of the codebase for a feature that is not essential, especially since the project is maintained by part time volunteers. This is what happened to pipenv; its complexity became overwhelming and the project suffered.  

At some point, users who want such specific features need to just make a venv and bash script instead of shoehorning it all into pipx. IMO the only way the suffix feature can work and pipx can remain viable is if the feature and implementation is kept extremely simple. I am fine not supporting suffixes at all, or only supporting them with functionality that is reduced but maintainable and easily documented. Maybe that's just me be being selfish because I'm perfectly happy with pipx as it is and don't want increased maintenance burden.

> * Only one suffix per venv
> * Only the main package and its apps get the suffix (not injected packages + their apps)

I'm fine with these limitations. 

> * Should `--include-deps` apps also get the suffix?

If `--include-deps` is used, and the same package was already installed with a different suffix, there will be name collisions if a suffix is not added to the apps of the dependent packages. So I think from a user perspective there has to be. I just wonder how hard this will be from a maintainer's perspective.

Maybe a simpler way to accomplish this and remove limitations is to symlink an entire folder with the suffix name, like `black_suffix` in the bin dir. Then users can run `black_suffix/black` from the command line.

**Comment by itsayellow** [comment]
What if we described suffixes as "experimental and subject to possible change" in both the help and docs?

**Comment by cs01** [comment]
> What if we described suffixes as "experimental and subject to possible change" in both the help and docs?

That sounds like a good idea since things are in flux.

**Comment by abn** [comment]
@cs01 just to add to the discussion, the suffix feature does not need to be complex. From my perspective, simply ensuring the main package is the only one that can be suffixed is good enough to reduce unnecessary complexity. 

The reason why I turned to `pipx` was this particular feature for what it is worth. This makes using multiple versions of python tools in the same environment a breeze. And I do sincerely hope that this feature remains, albeit with possible caveats.

**Comment by itsayellow** [comment]
> @cs01 just to add to the discussion, the suffix feature does not need to be complex. From my perspective, simply ensuring the main package is the only one that can be suffixed is good enough to reduce unnecessary complexity.
> 
> The reason why I turned to `pipx` was this particular feature for what it is worth. This makes using multiple versions of python tools in the same environment a breeze. And I do sincerely hope that this feature remains, albeit with possible caveats.

I think that's a good approach.

To summarize what @cs01 and I (and I believe @abn 🙂)  agree about:
* One suffix only per venv
* Suffix is applied to main package only
* Suffix is applied to apps of main package
* Suffix is applied to main package's dependencies' apps if `--include-deps` is used to install main package

Hopefully with this approach there is an implementation that keeps code complexity down.

**Comment by itsayellow** [comment]
As a minor side note:

> Maybe a simpler way to accomplish this and remove limitations is to symlink an entire folder with the suffix name, like `black_suffix` in the bin dir to the venv's bin dir. Then users can run `black_suffix/black` from the command line, and pipx doesn't have to do anything complicated other than create a venv and install to it.

Does this work?  I couldn't get `<subdirectory>/<app>` in a `PATH` directory to work for me in zsh.

**Comment by cs01** [comment]
> Does this work? I couldn't get `<subdirectory>/<app>` in a `PATH` directory to work for me in zsh.

I just tried it myself and it doesn't work. I assumed you could access folders on your PATH, but that doesn't seem to be the case.

File: src/pipx/commands/common.py
Comment (abn) [reply src/pipx/commands/common.py]: @itsayellow I have updated things a bit. The change defaults to venv's main package instead of venv root name. I have also removed the fallback loop to determine suffixed packaged.

Decided to retain the helper to abstract away some common functionality, and will also make maintenance easier in the future.

Let me know what you think about this version.

File: docs/changelog.md
Comment (itsayellow) [inline docs/changelog.md:4]: Can you change "Regression" to "Fixed regression".

File: tests/test_pipx_metadata_file.py
Comment (itsayellow) [inline tests/test_pipx_metadata_file.py:67]: Why is this necessary?

**Review by itsayellow — CHANGES_REQUESTED** [review]
Thanks, I think this is mostly ok, and a big improvement over our current broken implementation of suffixes.

I mainly have the one question about why you changed the name of the temp directory in `test_metadata_file_create()`.

File: tests/test_pipx_metadata_file.py
Comment (abn) [reply tests/test_pipx_metadata_file.py:67]: This is required because of https://github.com/pipxproject/pipx/pull/491/files#diff-349c0b5cf5479f95786c8426e2656d2bR105-R109

**Review by itsayellow — APPROVED** [review]

**Comment by joshbode** [comment]
Oops - I just noticed this PR - I made the initial suffix PR (#445) but I clearly didn't consider the impact enough (e.g. `pipx list`, etc).

Thank you for fixing this @abn :)

**Comment by abn** [comment]
@cs01 @itsayellow can we tag a new release with this fix in please? Seems like there isin't a new one yet.

**Pull request merged**
